### DISCUSSION TITLE
Firestore: `add()` does not support _WithConverterDocumentReference

### DISCUSSION THREAD

**Discussion by idkq** [firebase/flutterfire#6280] [Feature request]
When you `.add()` new Document and your collection was initiated with `.withConverter<Dog>()`, you'll get a `DocumentReference<Dog>` back as expected. Now, if you want to use this reference to add a **new second document that references the first doc**, it throws and exception.

```
Unhandled Exception: Invalid argument: Instance of '_WithConverterDocumentReference<Dog>'
#0      convertPlatformException (package:cloud_firestore_platform_interface/src/method_channel/utils/exception.dart:13:5)
#1      MethodChannelDocumentReference.set (package:cloud_firestore_platform_interface/src/method_channel/method_channel_document_reference.dart:43:13)
<asynchronous suspension>
#2      _JsonCollectionReference.add (package:cloud_firestore/src/collection_reference.dart:108:5)
<asynchronous suspension>
#3      _WithConverterCollectionReference.add (package:cloud_firestore/src/collection_reference.dart:179:22)
<asynchronous suspension>
```

Why? Because it expects a `_JsonDocumentReference` instead of a `_WithConverterDocumentReference`.

The solution is to get the document again from Firestore, but this time without the converter, and use the reference to add a new document.

There must be a better way to do this either by accessing _originalDocumentReference or having the `add()` method handling this automatically.

**Comment by rrousselGit** [comment]
Do you have some code samples to reproduce the issue?

**Reply by idkq** [reply]
```dart
Future<void> callThis() async {
    final modelsRef =
        FirebaseFirestore.instance.collection('models').withConverter<Model>(
              fromFirestore: (snapshot, _) => Model.fromJson(snapshot.data()!),
              toFirestore: (model, _) => model.toJson(),
            );

    final myDocRef = await modelsRef.add(Model());

    // Not allowed
    await modelsRef.add(Model(parent: myDocRef));
  }

class Model {
  final DocumentReference? parent;

  Model({this.parent});

  Model.fromJson(Map<String, dynamic> data) : parent = data["parent"];

  Map<String, dynamic> toJson() {
    return {"parent": parent};
  }
}
```

**Reply by rrousselGit** [reply]
This is not how withConverter works. You cannot have `toJson` return a Map that contains a DocumentReference. Nor will `data['parent']` will be a `DocumentReference` either

Maybe you want to do:

```dart
class Model {
  Model({required this.property});

  Model.fromJson(Map<String, dynamic> data) : property = data['property'];

  final String property;

  Map<String, dynamic> toJson() {
    return {'property': property};
  }
}
```

**Reply by idkq** [reply]
I guess you are confused with the example because of the variable parent. 

>  You cannot have toJson return a Map that contains a DocumentReference. Nor will data['parent'] will be a DocumentReference either

This is not what is expected here. I'm simply trying to add a new document and get a `DocumentReference` of the document added.

`var x = collection.add(Model())`

How can I use `x` as a `DocumentReference` in any other document, is the question. Think of it as a foreign key of two documents.
